File: src/application/usecase/RegisterIngredientUseCase.ts
import type { IdGenerator } from '@domain/contracts/IdGenerator.js'
import type { IngredientRepository } from '../contracts/IngredientRepository.js'
import type { RegisterIngredientCommand } from '../dto/command/RegisterIngredientCommand.js'
import type { IngredientDto } from '../dto/entity/IngredientDto.js'
import { IngredientAlreadyExistsException } from '../errors/IngredientAlreadyExistsException.js'
import { IngredientFactory } from '../factory/IngredientFactory.js'
import { IngredientMapper } from '../mapper/IngredientMapper.js'

export class RegisterIngredientUseCase {
  private readonly ingredientFactory: IngredientFactory

  constructor(
    private readonly ingredientRepository: IngredientRepository,
    private readonly idGenerator: IdGenerator
  ) {
    this.ingredientFactory = new IngredientFactory(this.idGenerator)
  }

  async execute(command: RegisterIngredientCommand): Promise<IngredientDto> {
    // Check if ingredient with the same name already exists
    const existingIngredient = await this.ingredientRepository.findByName(command.name)
    if (existingIngredient) {
      throw new IngredientAlreadyExistsException(command.name)
    }

    // Create new ingredient
    const ingredient = this.ingredientFactory.createFromCommand(command)

    // Save to repository
    await this.ingredientRepository.save(ingredient)

    // Return DTO
    return IngredientMapper.toDto(ingredient)
  }
}

File: api/src/main/java/br/unifor/costify/catalog/application/usecase/RegisterIngredientUseCase.java
package br.unifor.costify.catalog.application.usecase;

import br.unifor.costify.catalog.application.contracts.IngredientRepository;
import br.unifor.costify.catalog.application.dto.command.RegisterIngredientCommand;
import br.unifor.costify.catalog.application.dto.entity.IngredientDto;
import br.unifor.costify.catalog.application.errors.IngredientAlreadyExistsException;
import br.unifor.costify.catalog.application.factory.IngredientFactory;
import br.unifor.costify.shared.application.validation.ValidationService;
import br.unifor.costify.catalog.domain.entity.Ingredient;
import org.springframework.stereotype.Service;

@Service
public class RegisterIngredientUseCase {
  private final IngredientRepository ingredientRepository;
  private final IngredientFactory ingredientFactory;
  private final ValidationService validationService;

  public RegisterIngredientUseCase(
      IngredientRepository ingredientRepository, 
      IngredientFactory ingredientFactory,
      ValidationService validationService) {
    this.ingredientRepository = ingredientRepository;
    this.ingredientFactory = ingredientFactory;
    this.validationService = validationService;
  }

  public IngredientDto execute(RegisterIngredientCommand command) {
    validationService.validateIngredientData(
        command.name(), 
        command.packageQuantity(), 
        command.packagePrice(), 
        command.packageUnit()
    );
    
    validateIngredientDoesNotExist(command.name());

    Ingredient ingredient =
        ingredientFactory.create(
            command.name(),
            command.packageQuantity(),
            command.packagePrice(),
            command.packageUnit());

    Ingredient savedIngredient = ingredientRepository.save(ingredient);

    return IngredientDto.from(savedIngredient);
  }

  private void validateIngredientDoesNotExist(String name) {
    if (ingredientRepository.existsByName(name)) {
      throw new IngredientAlreadyExistsException(
          "Ingredient with name '" + name + "' already exists");
    }
  }
}
